src/braket/experimental/autoqasm/program/program.py

```diff
@@ -404,14 +423,31 @@ def get_expression_var(self, expression: FreeParameterExpression) -> oqpy.FloatV
         self.get_oqpy_program().declare(var)
         return var
 
-    def get_free_parameters(self) -> list[oqpy.FloatVar]:
+    def get_free_parameters(self) -> list[oqpy.Var]:
         """Return a list of named oqpy.Vars that are used as free parameters in the program."""
         return list(self._free_parameters.values())
 
+    def get_free_parameter(self, name: str) -> oqpy.Var | None:
+        """Return the oqpy.Var associated with the variable name `name` in the program."""
+        return self._free_parameters.get(name, None)
+
+    def get_output_parameter(self, name: str) -> oqpy.Var | None:
+        """Return the oqpy.Var associated with the output variable `name` in the program."""
+        return self._output_parameters.get(name, None)
+
     def add_io_declarations(self) -> None:
         """Add input and output declaration statements to the program."""
         root_oqpy_program = self.get_oqpy_program(scope=ProgramScope.MAIN)
-        for parameter in self.get_free_parameters():
+        for parameter_name, parameter in self._free_parameters.items():
+            # Sometimes the names get overwritten -- see aq.types
+            parameter.name = parameter_name
+            root_oqpy_program._add_var(parameter)
+        for parameter_name, parameter in self._output_parameters.items():
+            parameter.name = parameter_name
+            if parameter_name in root_oqpy_program.undeclared_vars:
+                del root_oqpy_program.undeclared_vars[parameter_name]
+            if parameter_name in root_oqpy_program.declared_vars:
+                del root_oqpy_program.declared_vars[parameter_name]
```

Comment:
Maybe a slight readability improvement, and an explanatory comment (feel free to re-word this):
```suggestion
        for parameter_name, parameter in self._output_parameters.items():
            # Before adding the output variable to the program, remove any existing reference
            root_oqpy_program.undeclared_vars.pop(parameter_name, None)
            root_oqpy_program.declared_vars.pop(parameter_name, None)
            
            parameter.name = parameter_name
```